File: Semester 5/Mobile app programming/Pawtect/app/src/main/java/com/example/pawtect/ui/screen/HomeScreen.kt
package com.example.pawtect.ui.screen

import PetViewModel
import android.annotation.SuppressLint
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material3.BottomAppBar
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FabPosition
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.example.pawtect.R
import com.example.pawtect.data.PetData
import com.example.pawtect.data.PetEntity
import com.example.pawtect.data.PetRepository
import com.example.pawtect.ui.theme.DeepBlue
import com.example.pawtect.ui.theme.LightBlue
import com.example.pawtect.ui.theme.LighterBlue

@SuppressLint("UnusedMaterial3ScaffoldPaddingParameter")
@Composable
fun HomeScreen(navController: NavController, viewModel: PetViewModel)  {
    val petList by viewModel.readAllData.observeAsState(emptyList())

    DisposableEffect(Unit) {
        onDispose {  }
    }

    Scaffold(
        bottomBar = {
            BottomAppBar(
                contentPadding = PaddingValues(16.dp),
                containerColor = Color.Transparent
            ) {}
        },
        floatingActionButtonPosition = FabPosition.End,
        floatingActionButton = {
            FloatingActionButton(onClick = {
                navController.navigate("add")
            }, containerColor = LighterBlue, shape = CircleShape) {
                Icon(
                    imageVector = Icons.Filled.Add,
                    contentDescription = "Add",
                    tint = Color.White,
                )
            }
        }) {
        Box(
            modifier = Modifier
                .background(DeepBlue)
                .padding(24.dp)
                .fillMaxSize()
        ) {
            LazyColumn() {
                item {
                    Header()
                    Spacer(modifier = Modifier.height(24.dp))
                    SearchField()
                    Spacer(modifier = Modifier.height(24.dp))
                    Text(
                        "New pets",
                        style = TextStyle(color = Color.White, fontWeight = FontWeight.Bold)
                    )
                    Spacer(Modifier.height(12.dp))
                }
                items(petList) { petEntity ->
                    PetListItem(item = petEntity, onCardClick = { pet ->
                        navController.navigate("details/${pet.id}")
                    })
                }
            }
        }
    }
}

@Composable
fun Header(name: String = "pet adopter") {
    Row(
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.fillMaxWidth()
    ) {
        Column(verticalArrangement = Arrangement.Center) {
            Text(
                text = "Welcome, $name",
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold,
                fontSize = 30.sp,
                color = Color.White
            )
            Text(
                text = "Hoping a pet is finding an owner and that will be you",
                style = MaterialTheme.typography.bodyMedium,
                color = Color.White
            )
        }
        Icon(
            painter = painterResource(id = R.drawable.baseline_pets_24),
            contentDescription = "  a pet icon",
            tint = Color.White,
            modifier = Modifier
                .size(45.dp)
                .rotate(45.0f)
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SearchField() {
    val textFieldValue = remember { mutableStateOf("") }
    TextField(
        value = textFieldValue.value,
        onValueChange = { value -> textFieldValue.value = value },
        textStyle = TextStyle(
            fontSize = 12.sp,
            color = Color.White,
        ),
        maxLines = 1,
        placeholder = {
            Text(
                text = "Write a pet name",
                color = Color.White,

                )
        },
        leadingIcon = {
            Icon(
                painter = painterResource(id = R.drawable.baseline_search_24),
                contentDescription = "Search icon button",
                tint = Color.White
            )
        },
        colors = TextFieldDefaults.textFieldColors(
            focusedIndicatorColor = Color.Transparent,
            disabledIndicatorColor = Color.Transparent,
            errorIndicatorColor = Color.Transparent,
            unfocusedIndicatorColor = Color.Transparent,
        ),
        modifier = Modifier.fillMaxWidth(),
    )
}

@Composable
fun PetListItem(item:PetEntity, onCardClick: (PetEntity) -> Unit){
    Card(
        modifier = Modifier
            .padding(bottom = 16.dp)
            .clip(RoundedCornerShape(size = 8.dp))
            .background(LightBlue, RoundedCornerShape(8.dp))
            .clickable(onClick = { onCardClick(item) }, enabled = true)
            .padding(12.dp),
        elevation = CardDefaults.cardElevation(0.dp),
        shape = RoundedCornerShape(8.dp),
    ) {
        Column(Modifier.padding(8.dp)) {
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(200.dp)
                    .clip(shape = RoundedCornerShape(6.dp))
                    .background(color = Color.Gray),
                Alignment.Center
            ) {
                Image(
                    painter = painterResource(id = item.imagePath),
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.fillMaxSize()
                )
            }
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                item.name,
                style = TextStyle(
                    color =  Color.White,
                    fontWeight = FontWeight.Bold,
                    fontSize = 22.sp
                )
            )
            PetDetails(age = item.age, weight = item.weight, species = item.species)
        }
    }
}

@Composable
fun PetDetails(age: Int, weight:Double, species:String){
    Row(horizontalArrangement = Arrangement.SpaceBetween){
        Column(
            Modifier
                .weight(1F)
                .align(Alignment.CenterVertically)
        ){
            val year = if (age > 1) "Years" else "Year"
            Text("Age", style = TextStyle(
                color = Color.White,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold
            ),
                textAlign = TextAlign.Center
                )
            Text("$age $year", textAlign = TextAlign.Center, color =  Color.White)
        }
        Column(
            Modifier
                .weight(1F)
                .align(Alignment.CenterVertically)
        ){
            val kg = if (weight > 1) "kgs" else "kg"
            Text("Weight", style = TextStyle(
                color = Color.White,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold
            ),
                textAlign = TextAlign.Center
            )
            Text("$weight $kg", textAlign = TextAlign.Center, color = Color.White)
        }
        Column(
            Modifier
                .weight(1F)
                .align(Alignment.CenterVertically)
        ){
            Text("Species", style = TextStyle(
                color = Color.White,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold
            ),
                textAlign = TextAlign.Center
            )
            Text("$species", textAlign = TextAlign.Right, color =  Color.White)
        }
    }
}

File: Semester 5/Mobile app programming/pawtect_flutter/lib/home_screen.dart
import 'package:flutter/material.dart';
import 'package:pawtect_flutter/pet_model.dart';
import 'package:pawtect_flutter/pet_repository.dart';

class HomeScreen extends StatelessWidget {
  const HomeScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButtonLocation: FloatingActionButtonLocation.endDocked,
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          Navigator.pushNamed(context, '/add');
        },
        backgroundColor: Color(0xFF5d9dc4),
        child: const Icon(
          Icons.add,
          color: Colors.white,
        ),
      ),
      body: SafeArea(
        child: Container(
          decoration: BoxDecoration(
            color: Color(0xFF072233),
          ),
          padding: const EdgeInsets.fromLTRB(24.0, 32.0, 24.0, 24.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Header(),
              const SizedBox(height: 24.0),
              SearchField(),
              const SizedBox(height: 24.0),
              const Text(
                "New pets",
                style: TextStyle(color: Colors.white, fontWeight: FontWeight.bold),
              ),
              const SizedBox(height: 12.0),
              Expanded(
                child: FutureBuilder<List<Pet>>(
                  future: PetRepository.getAllPets(),
                  builder: (context, snapshot) {
                    if (snapshot.connectionState == ConnectionState.waiting) {
                      return CircularProgressIndicator();
                    } else if (snapshot.hasError) {
                      return Text('Error: ${snapshot.error}');
                    } else if (!snapshot.hasData || snapshot.data!.isEmpty) {
                      return Text('No pets available.');
                    } else {
                      return ListView.builder(
                        itemCount: snapshot.data!.length, // to look into it further
                        itemBuilder: (context, index) {
                          return PetListItem(item: snapshot.data![index], onCardClick: (pet) {
                          });
                        },
                      );
                    }
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


class Header extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              "Welcome, pet adopter",
              style: const TextStyle(
                fontWeight: FontWeight.bold,
                fontSize: 30.0,
                color: Colors.white,
              ),
            ),
            const Text(
              "Hoping a pet is finding an owner and that will be you",
              style: TextStyle(color: Colors.white),
            ),
          ],
        ),
        Transform.rotate(
          angle: 45.0,
          child: const Icon(
            Icons.pets,
            size: 45.0,
            color: Colors.white,
          ),
        ),
      ],
    );
  }
}

class SearchField extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return TextField(
      style: const TextStyle(fontSize: 12.0, color: Colors.white),
      decoration: InputDecoration(
        hintText: "Write a pet name",
        hintStyle: const TextStyle(color: Colors.white),
        prefixIcon: const Icon(Icons.search, color: Colors.white),
      ),
    );
  }
}

class PetListItem extends StatelessWidget {
  final Pet item;
  final void Function(Pet) onCardClick;

  const PetListItem({Key? key, required this.item, required this.onCardClick}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Card(
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8.0)),
      color: Colors.black12,
      elevation: 0.0,
      margin: const EdgeInsets.only(bottom: 16.0),
      child: InkWell(
        onTap: () {
          onCardClick(item);
          Navigator.pushNamed(context, '/details', arguments: {'petId': item.id});
        },
        child: Padding(
          padding: const EdgeInsets.all(12.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Container(
                height: 200.0,
                decoration: BoxDecoration(
                  color: Colors.grey,
                  borderRadius: BorderRadius.circular(6.0),
                ),
                child: Image.asset(
                  item.imagePath,
                  fit: BoxFit.cover,
                ),
              ),
              const SizedBox(height: 4.0),
              Text(
                item.name,
                style: const TextStyle(
                  color: Colors.white,
                  fontWeight: FontWeight.bold,
                  fontSize: 22.0,
                ),
              ),
              PetDetails(age: item.age, weight: item.weight, species: item.species),
            ],
          ),
        ),
      ),
    );
  }
}


class PetDetails extends StatelessWidget {
  final int age;
  final double weight;
  final String species;

  const PetDetails({Key? key, required this.age, required this.weight, required this.species}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        _DetailColumn(label: "Age", value: "$age ${age > 1 ? 'Years' : 'Year'}"),
        _DetailColumn(label: "Weight", value: "$weight ${weight > 1 ? 'kgs' : 'kg'}"),
        _DetailColumn(label: "Species", value: species),
      ],
    );
  }
}

class _DetailColumn extends StatelessWidget {
  final String label;
  final String value;

  const _DetailColumn({Key? key, required this.label, required this.value}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
      Text(
      label,
      style: const TextStyle(
        color: Colors.white,
        fontSize: 18.0,
        fontWeight: FontWeight.bold,
      ),
    ),
    Text(
    value,
    style
        : const TextStyle(
      color: Colors.white,
    ),
    ),
      ],
    );
  }
}
